AutomatorServer/android/app/src/androidTest/java/pl/leancode/automatorserver/PatrolAutomator.kt

```diff
@@ -334,71 +334,88 @@ class PatrolAutomator private constructor() {
-        val identifiers = listOf(
+        val identifiers = arrayOf(
             "com.android.packageinstaller:id/permission_allow_button",
             "com.android.permissioncontroller:id/permission_allow_button",
             "com.android.permissioncontroller:id/permission_allow_one_time_button",
         )
 
-        for (identifier in identifiers) {
-            if (exists(identifier)) {
-                uiDevice.findObject(UiSelector().resourceId(identifier)).click()
-                return
-            }
-        }
+        val uiObject = waitForUiObjectByResourceId(*identifiers)
+            ?: throw UiObjectNotFoundException("button to allow permission once")
 
-        throw UiObjectNotFoundException("button to allow permission once")
+        uiObject.click()
     }
 
     fun denyPermission() {
-        val identifiers = listOf(
+        val identifiers = arrayOf(
             "com.android.packageinstaller:id/permission_deny_button",
             "com.android.permissioncontroller:id/permission_deny_button",
         )
 
-        for (identifier in identifiers) {
-            if (exists(identifier)) {
-                uiDevice.findObject(UiSelector().resourceId(identifier)).click()
-                return
-            }
-        }
+        val uiObject = waitForUiObjectByResourceId(*identifiers)
+            ?: throw UiObjectNotFoundException("button to deny permission")
 
-        throw UiObjectNotFoundException("button to deny permission")
+        uiObject.click()
     }
 
     fun selectFineLocation() {
         val resourceId = "com.android.permissioncontroller:id/permission_location_accuracy_radio_fine"
-        if (uiDevice.findObjects(By.res(resourceId)).isNotEmpty()) {
-            uiDevice.findObject(UiSelector().resourceId(resourceId)).click()
-        } else {
-            throw UiObjectNotFoundException("button to select fine location")
-        }
+
+        val uiObject = waitForUiObjectByResourceId(resourceId)
+            ?: throw UiObjectNotFoundException("button to select fine location")
+
+        uiObject.click()
     }
 
     fun selectCoarseLocation() {
         val resourceId = "com.android.permissioncontroller:id/permission_location_accuracy_radio_coarse"
-        if (uiDevice.findObjects(By.res(resourceId)).isNotEmpty()) {
-            uiDevice.findObject(UiSelector().resourceId(resourceId)).click()
-        } else {
-            throw UiObjectNotFoundException("button to select coarse location")
+
+        val uiObject = waitForUiObjectByResourceId(resourceId)
+            ?: throw UiObjectNotFoundException("button to select coarse location")
+
+        uiObject.click()
+    }
+
+    private fun waitForSelector(bySelector: BySelector) {
+        val startTime = System.currentTimeMillis()
+        while (System.currentTimeMillis() - startTime < timeout) {
+            if (uiDevice.findObjects(bySelector).isNotEmpty()) {
+                return
+            }
+
+            delay(ms = 500)
```

Comment:
Try refactoring this using `launch {}` (this will probably leave a `while (true)` loop though, so IDK if it's any better)

Reply:
I tried to use fancy coroutines but unfortunately it just turned out to be a total mess.

This is good enough and very similar to the iOS Swift implementation.